beard/clustering/blocking.py

```diff
@@ -159,33 +183,48 @@ def _blocks(self, X, y, blocks):
             if self.affinity == "precomputed":
                 X_mask = X_mask[:, mask]
 
-            # Select a clusterer
-            if len(X_mask) == 1:
-                clusterer = _SingleClustering()
-            elif self.fit_:
-                # Although every job has its own copy of the estimator, one job
-                # can serve multiple fits. That's why the clone is needed.
-                clusterer = clone(self.base_estimator)
-            elif self.partial_fit_:
-                if b in self.clusterers_:
-                    clusterer = self.clusterers_[b]
-                else:
-                    clusterer = clone(self.base_estimator)
-
-            yield (b, X_mask, y_mask, clusterer)
+            yield (b, X_mask, y_mask)
 
     def _fit(self, X, y, blocks):
         """Fit base clustering estimators on X."""
         self.blocks_ = blocks
 
-        results = (Parallel(n_jobs=self.n_jobs, verbose=self.verbose)
-                   (delayed(_parallel_fit)(self.fit_, self.partial_fit_,
-                                           b, X_mask, y_mask, clusterer,
-                                           self.verbose) for
-                   b, X_mask, y_mask, clusterer in self._blocks(X, y, blocks)))
-
-        for b, clusterer in results:
-            self.clusterers_[b] = clusterer
+        processes = []
+        # Here the blocks will be passed to subprocesses
+        data_queue = mp.Queue()
+        # Here the results will be passed back
+        result_queue = mp.Queue()
+        for x in range(0, self.n_jobs):
+            processes.append(mp.Process(target=_parallel_fit, args=(self.fit_,
+                             self.partial_fit_, self.base_estimator,
+                             self.verbose, data_queue, result_queue)))
+            processes[-1].start()
+
+        # First 16 blocks are sent into the queue without waiting for the
+        # results. This variable is a counter that takes care of this.
+        presend = 1
+
+        for block in self._blocks(X, y, blocks):
+            if presend > self.n_jobs:
```

Comment:
From a rigorous loop invariant point of view, I would initialize `presend` to `0` and then check whether `presend >= self.n_jobs` instead. Or better, I would put this check after the `data_queue.put` call below (using a `>` then).